spec outline

tzumainn opened the pull request.

tzumainn committed:
spec outline

larsks commented on specs/backlog/access-control.rst:
```diff
@@ -0,0 +1,189 @@
+Right now Ironic is not multi-tenant; anyone with API access to one node
+has access to all nodes. However there is demand for providing finer-grained
+access control to nodes. This spec describes a solution that uses a custom
+policy engine to do so.
```
Saying "anyone with API access to one node has access to all nodes" suggests that it is possible for a non-admin user to meaningfully interact with Ironic, which I think is not the case. I would say, "Ironic only permits users with admin privileges to access nodes...", etc.

tzumainn replied:
This is true though, isn't it? If someone has Ironic API access to one node, that means that they have Ironic API access to all nodes. That's the fundamental problem we're trying to solve, right?

larsks commented on specs/backlog/access-control.rst:
```diff
@@ -0,0 +1,189 @@
+Problem description
+===================
+
+There are a variety of use cases where this capability is needed:
```
There should be something *before* this that describes the problem (or just restructure this so we don't immediately jump into bullet points, which are typically used to support a previously expressed idea).

tzumainn replied:
Fair point; I'll add some text here - a short rephrasing of the introduction.

larsks commented on specs/backlog/access-control.rst:
```diff
@@ -0,0 +1,189 @@
+  Each Owner needs to be able to manage their own nodes through the Ironic
+  API, while being unable to manage others' nodes.
+
+* An Ironic deployment wishes to grant specific Users the capability of
```
I don't think an Ironic deployment can wish anything. Maybe "an Openstack operator"?

tzumainn replied:
Did you ever ask one? I'll make the change though :)

larsks commented on specs/backlog/access-control.rst:
```diff
@@ -0,0 +1,189 @@
+a policy engine that is used throughout the REST API. We propose creating an
+extended version of this policy engine for use solely by the Node API. This
+policy engine still relies upon the Oslo policy file, but also enables the
+following access:
```
When we say "the Node API", to what exactly are we referring? We should make that explicit in the spec.

tzumainn replied:
I'll clarify to "Ironic node API"

larsks commented on specs/backlog/access-control.rst:
```diff
@@ -0,0 +1,189 @@
+* Allows an Owner of a node to automatically pass any policy check if they
+  own the node.
+
+* Allows a User of a node to pass a policy check if they have the appropriate
```
What is a "User of a node"? And we haven't really defined Owner, either. Maybe prefix this with some definitions?

tzumainn replied:
I'll put a brief description of User and Owner at the beginning of the spec.

larsks commented on specs/backlog/access-control.rst:
```diff
@@ -0,0 +1,189 @@
+
+The latter may be expanded to specify a policy set - another table that maps
+to a list of policies - rather than a single policy. This enables easy reuse
+of a group of policies.
```
I think we should handle authorization in a manner analogous to how Nova handles port based access control: you never open ports directly on an instance; you *always* use a security group. Similarly, I think all authorization should be defined in terms of our policy set idea.

tzumainn replied:
Yeah, I think I structured this section badly. I'll rework it.

larsks commented on specs/backlog/access-control.rst:
```diff
@@ -0,0 +1,189 @@
+
+This alternative policy engine is enough to control access for most API
+functions - except for list functions. For those, we propose updating the
+node list query to filter the result for non-admins.
```
"...we propose updating the node list query to behave similar to other multi-tenant openstack services: a user without admin privileges will only be able to list their own nodes, while an admin user will continue to see nodes for all users"

I think it's key to emphasize that our goal is *no change* to the default behavior of Ironic.

tzumainn replied:
Yep, excellent point; if this node policy engine is used, and the owner is not set, then Ironic's behavior will be the same. I'll emphasize that.

larsks commented on specs/backlog/access-control.rst:
```diff
@@ -0,0 +1,189 @@
+node list query to filter the result for non-admins.
+
+A switch in the configuration would control whether you're using the "classic"
+Ironic policy engine versus this new fine-grained policy engine.
```
Do you think this is necessary? Assuming we're agreed on the "no change to default behavior", then couldn't we use the new policy engine in all cases?

tzumainn replied:
Well, the problem is that although the `owner` field is intended to be the owner's project, it could theoretically be anything. I think there could be unexpected behavior if someone is using Ironic and sets the owner field to a non-project_id value.

tzumainn commented:
I've been poking around some other OpenStack code, and slooooowly realized something, which is as follows:

* here's an ironic policy check: https://github.com/openstack/ironic/blob/master/ironic/api/controllers/v1/node.py#L225 . It passes in `cdict` twice, but the first one gets assigned to a `target` variable in the `authorize` function

* after digging into the way some other OpenStack services use the `target` variable, I realized it's supposed to represent the object you're doing the check on, which made me wonder why; how can it be used?

* then I found this blog: http://jaormx.github.io/2018/oslo-policy-deep-dive-p1/ - check out the section about 'Target object attributes' (this is undocumented in the official oslo.policy documentation, incidentally)

What this means is I think we can accomplish what we want - perhaps to a more limited degree, but with the advantage of not creating any database tables - by:

* adding a `users` list field to nodes, which is simply list of User projects that can access the node
* updating the node controller to pass in information about the node to the policy check, including the owner and whether the current user is in the `users` list (I'm not sure that the policy file can check whether a project_id is in a list)
* documenting the proper settings to the oslo policy file

And that would be it!

I'll dig some more, but I may rework the spec if I think it looks promising.

tzumainn commented:
Oh, and this would leave us with far less fine-grained control over what individual users can do, but there's still differentiation between owners and users - and maybe that's enough?

tzumainn commented:
The default nova policy file already does this a little: https://docs.openstack.org/nova/latest/configuration/sample-policy.html

I think we would create two - policy roles? - called node_owner and node_user that specifically check target.node.owner and target.node.user; and then allow the Ironic admin to update the policy file as they see fit. Maybe set certain defaults?

tzumainn committed:
add alternative node policy spec

tzumainn commented:
I've added an alternative spec - let me know what you think!

tzumainn committed:
add node-owner policy spec

larsks commented on specs/backlog/node-owner-policy.rst:
```diff
@@ -0,0 +1,195 @@
+is inconvenient if an Ironic deployment consists of nodes from multiple
+hardware Owners: each Owner needs to be able to manage their own nodes
+through the Ironic API, while being unable to manage others' nodes.
+
```
I would like to make the problem statement a little stronger (e.g., it's more than inconvenient; the lack of multi-tenancy actively makes certain uses of Ironic impossible). I'll muddle over the phrasing a little bit this afternoon.

tzumainn replied:
I've made adjustments, let me know what you think!

larsks commented on specs/backlog/node-owner-policy.rst:
```diff
@@ -0,0 +1,195 @@
+
+This update is enough to control access for most API functions - except for list
+functions. For those, we propose updating the node list query so that non-admin
+results are filtered against the ``owner`` field.
```
"...in a fashion analogous to other openstack services."

tzumainn replied:
Added.

larsks commented on specs/backlog/node-owner-policy.rst:
```diff
@@ -0,0 +1,195 @@
+Other end user impact
+---------------------
+
+None
```
I think that when our policy engine is active, we should validate the value of the `owner` field when it is set or updated. Would this count as a REST API change?

tzumainn replied:
Does that mean we should have a configuration setting so we know when it's active?

I'm also not sure it's necessary from an OpenStack-pattern point of view.

larsks commented on specs/backlog/node-owner-policy.rst:
```diff
@@ -0,0 +1,195 @@
+controller uses the policy engine[1], it does so without passing in any
+information about the node being checked. Other services such as Nova[2] and
+Barbican[3] pass in additional information, and we propose doing the same. In
+summary, we would like to:
```
Nitpicks: we should fix the rst syntax here (and throughout) so that rather than footnotes we have active links.

tzumainn replied:
From what I've seen of other specs, they seem to prefer footnotes - maybe there's rst-specifc syntax for that I don't know about ... ?

larsks committed:
Implement multi-tenancy via the "owner" field.

This spec proposes repurposing the `owner` field for use as an access
control mechanism to implement multi-tenant behavior in Ironic.

Story: 2006506
Task: 36486

tzumainn committed:
Merge pull request #2 from larsks/spec/node-owner

Implement multi-tenancy via the "owner" field.
